### ISSUE TITLE
functional.pad returns original tensor when padding size is 0

### ISSUE THREAD

**Issue by ppwwyyxx** [pytorch/pytorch#31734]
## To Reproduce

```python
import torch
from torch.nn import functional as F
x = torch.ones(3, 10)

y1 = F.pad(x, [1, 1, 1, 1])
y2 = F.pad(x, [0, 0, 0, 0])
y3 = x + 0

y1.unsqueeze_(0)
print(x.shape)  # 3, 10

y3.unsqueeze_(0)
print(x.shape)  # 3, 10

y2.unsqueeze_(0)
print(x.shape)  # 1, 3, 10
print(x is y2)  # True

y1.requires_grad = True
print(x.requires_grad)  # False
y3.requires_grad = True
print(x.requires_grad)  # False
y2.requires_grad = True
print(x.requires_grad)  # True

```

## Expected behavior

I expect y2 to be a different tensor from x, although I understand this may be arguable.

In the current situation, the semantic of `F.pad` is different depending the padded #pixels: its output may be the same input tensor sometimes. And modifying the output tensor will __sometimes__ modify the input tensor - potentially lead to unnoticed dangerous consequences. 
As a related note, `x + 0` and `x[:, :]` is not the same tensor as `x` and I think doing so would be a bad idea.

To avoid such dangerous consequences, currently one will need to handle padding_size=0 manually.

cc @ezyang @gchanan @zou3519

**Comment by One-sixth** [comment]
I agree with you that this also caused my custom backpropagation function to fail in my code. #31748

**Comment by ezyang** [comment]
I agree with the diagnosis here.

**Comment by ezyang** [comment]
Some other cases to be careful about: we think maybe dropout with 0 returns original tensor, and pow with 1 may also return the original tensor (needs double checking).

**Comment by ezyang** [comment]
We discussed this in triage review and we agree that we should return a clone here.
